### ISSUE TITLE
[FileSearch] Support wildcard characters “@“ “#"

### ISSUE THREAD

**Issue by jasonLaster** [devtools-html/debugger.html#2446]
It would be really cool if we could support wildcard characters at the beginning of queries like "@" and "#". The UX is like:

1. type "@"
 = > we select function search mode and highlight the button
2. type "on" 
  => we search for functions starting with "on"
  
#### diff

* doesn't cover firing an action to change the mode when the wildcard changes

```diff
diff --git a/src/components/Editor/SearchBar.js b/src/components/Editor/SearchBar.js
index b82e267..5b86a0d 100644
--- a/src/components/Editor/SearchBar.js
+++ b/src/components/Editor/SearchBar.js
@@ -405,7 +405,18 @@ const SearchBar = React.createClass({
   },

   async onChange(e: any) {
-    return this.doSearch(e.target.value);
+    let query = e.target.value;
+
+    // check the first character of the query for special characters...
+    if (hasWildCard(query)) {
+      const firstChar = query[0]
+      query = query.slice(1, -1);
+
+      //....
+    }
+
+
+    return this.doSearch(query);
   },
```

**Comment by jacobjzhang** [comment]
@jasonLaster If I'm understanding this correctly, the @ character activates function search?

When I think wildcard, I think a search term like "*ey" will look for terms like "key", "they", etc.

**Comment by wldcordeiro** [comment]
> When I think wildcard, I think a search term like "*ey" will look for terms like "key", "they", etc.

I think having Regex search beats that 😄

**Comment by AnshulMalik** [comment]
Can we do something like this ?
```diff
diff --git a/src/components/Editor/SearchBar.js b/src/components/Editor/SearchBar.js
index b16ee37..4ffde22 100644
--- a/src/components/Editor/SearchBar.js
+++ b/src/components/Editor/SearchBar.js
@@ -473,7 +473,13 @@ class SearchBar extends Component {
   }
 
   async onChange(e: any) {
-    return this.doSearch(e.target.value);
+    let query = e.target.value;
+
+    if (query[0] == "@") {
+      query = query.slice(1);
+      this.toggleSymbolSearch(e, { toggle: false, searchType: "functions" });
+    }
+    return this.doSearch(query);
   }
 
   onKeyUp(e: SyntheticKeyboardEvent) {
```

**Comment by jasonLaster** [comment]
yeah! that looks good. give it a try!

**Comment by jasonLaster** [comment]
@AnshulMalik had a great implementation, but this is currently blocked on the search modal and the UX we want afterwards. Closing his temporarily.

**Comment by jasonLaster** [comment]
Moving this to the enhancements board as it could be complicated and is a lower priority than core UI features:

https://github.com/devtools-html/debugger.html/projects/5
